scripts/train_lm.py

```diff
@@ -146,12 +146,39 @@ def WriteMetaparameters(metaparameters, ngram_order, num_train_sets, out_file):
 
 def FormatMetaparameters(metaparameters):
     out = []
+    # get the maximum length of decimals of metaparameters
+    decimal_length = 0
     for param in metaparameters:
-        x = '{:.3f}'.format(param)
-        if x == '0.00':
-            x = '{:.3g}'.format(param)
-        out.append(x)
-
+        temp_length = len(str(param))
+        if decimal_length < temp_length:
+            decimal_length = temp_length
+    decimal_length -= 2
+    
+    temp_original = metaparameters[0]
+    temp = round(temp_original, 3)
+    for param in metaparameters[1:]:
+        round_index = 3
+        x_original = param
+        x = round(x_original, round_index)
+        while(temp == x and round_index <= decimal_length):
```

Comment:
Hi, Dan, Jian
Ke discussed this issue with me, and I think there can be a very tricky way of doing this,

def FormatMetaparameters(metaparameters, num_train_sets):
    out = []
    for num_train_set in range(num_train_sets):
        count_scale = str(metaparameters[0])[:5]
        # note count_scale will be valid if input is valid (i.e. 0 and 1 will not be generated after approximaing in this way)
        out.append(count_scale)
    for order_minus_two in range( (len(metaparameters) - 1) / 4): 
        for D_idx in range(1,5):
            out.append( str(metaparameters[order_minus_two * 4 + D_idx])[:5] + str(5 - D_idx) )
            # add a extra digit to satisfy conditions after we rounded in this way
    return ','.join(out)

For example, if metaparameters = [1e-12, 0.123123123123123, 0.1231, 0.12301, 0.1230001, 0.56456456, 0.4356564, 0.1232323, 1e-08]
the output is FormatMetaparameters(metaparameters,1) = '1e-12,0.1234,0.1233,0.1232,0.1231,0.5644,0.4353,0.1232,1e-081’

I know there can be some weird thing happen in extreme rare cases, but I believe this can rarely happen in practice (maybe?)

best regards,
Zhouyang


On Aug 16, 2016, at 6:30 PM, Daniel Povey <notifications@github.com> wrote:

> In scripts/train_lm.py:
> 
> > -        if x == '0.00':
> > -            x = '{:.3g}'.format(param)
> > -        out.append(x)
> > -
> > +        temp_length = len(str(param))
> > +        if decimal_length < temp_length:
> > +            decimal_length = temp_length
> > +    decimal_length -= 2
> > +    
> > +    temp_original = metaparameters[0]
> > +    temp = round(temp_original, 3)
> > +    for param in metaparameters[1:]:
> > +        round_index = 3
> > +        x_original = param
> > +        x = round(x_original, round_index)
> > +        while(temp == x and round_index <= decimal_length):
> I don't think you understood what I was saying.
> The idea is to choose the decimal length independently for each metaparameter, in such a way that
> 
> we don't get exact ones or zeros
> we don't get collisions producing identical metaparameters. Plus, the code needs to be much clearer. Use variable names more carefully.
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>